src/config/reader.rs

```diff
@@ -0,0 +1,227 @@
+      return Err(anyhow!("Read over URL failed with status code: {}", resp.status()));
+    }
+    let source = if let Some(v) = resp.headers().get("content-type") {
+      if let Ok(s) = Source::detect_content_type(v.to_str()?) {
+        s
+      } else {
+        Source::detect(path.trim_end_matches('/'))?
+      }
+    } else {
+      Source::detect(path.trim_end_matches('/'))?
+    };
+    let txt = resp.text().await?;
+    Ok((txt, source))
+  }
+  async fn from_url(url: Url) -> anyhow::Result<Config> {
+    let (st, source) = Self::read_over_url(url).await?;
+    Config::from_source(source, &st)
+  }
+}
+#[cfg(test)]
+mod reader_tests {
+  use crate::config::reader::ConfigReader;
+
+  static TEST_GQL_BODY: &str = r#"
+        schema
+        @server(port: 8000, enableGraphiql: true, enableQueryValidation: false, hostname: "0.0.0.0")
+        @upstream(baseURL: "http://jsonplaceholder.typicode.com", enableHttpCache: true) {
+        query: Test
+      }
+      type Test {
+        id: Int!
+        name: String!
+        username: String!
+        email: String!
+        phone: String
+        website: String
+      }
+  "#;
+
+  static TEST_JSON_BODY: &str = r#"
+  {
+  "server": {
+    "port": 8000,
+    "enableGraphiql": true,
+    "enableQueryValidation": false,
+    "hostname": "0.0.0.0"
+  },
+  "upstream": {
+    "baseURL": "http://jsonplaceholder.typicode.com",
+    "enableHttpCache": true
+  },
+  "graphql": {
+    "schema": {
+      "query": "Query"
+    },
+    "types": {
+      "Post": {
+        "fields": {
+          "body": {
+            "type_of": "String",
+            "required": true
+          },
+          "id": {
+            "type_of": "Int",
+            "required": true
+          },
+          "title": {
+            "type_of": "String",
+            "required": true
+          },
+          "user": {
+            "type_of": "User",
+            "http": {
+              "path": "/users/{{value.userId}}"
+            }
+          },
+          "userId": {
+            "type_of": "Int",
+            "required": true
+          }
+        }
+      },
+      "Query": {
+        "fields": {
+          "posts": {
+            "type_of": "Post",
+            "list": true,
+            "http": {
+              "path": "/posts"
+            }
+          }
+        }
+      },
+      "User": {
+        "fields": {
+          "email": {
+            "type_of": "String",
+            "required": true
+          },
+          "id": {
+            "type_of": "Int",
+            "required": true
+          },
+          "name": {
+            "type_of": "String",
+            "required": true
+          },
+          "phone": {
+            "type_of": "String"
+          },
+          "username": {
+            "type_of": "String",
+            "required": true
+          },
+          "website": {
+            "type_of": "String"
+          }
+        }
+      }
+    },
+    "unions": {}
+  }
+}
+  "#;
```

Comment:
Instead of writing a config as string u could create one using Config:: default () and then convert it to sdl and other formats. There is an API to convert config to yaml json and graphql format.